My Project/clsGenerateTranslations.vb

```diff
@@ -0,0 +1,580 @@
+        End Try
+
+        If iRowsUpdated <> clsDatatableTranslations.Rows.Count Then
+            Throw New Exception("Error: Could NOT save all form id texts to the translations table. Rows saved: " & iRowsUpdated)
+        End If
+
+        Return iRowsUpdated
+    End Function
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>
+    ''' Updates the 'translations' database table with the texts from controls in the passed datatable
+    ''' </summary>   
+    ''' <param name="strDataSource">The file path to the sqlite database</param>
+    ''' <param name="clsDatatableControls">The form controls datatable with 3 columns; form_name, control_name, id_text.</param>
+    ''' <returns>The number of successful updates.</returns>
+    '''--------------------------------------------------------------------------------------------
+    Public Shared Function UpdateTranslationsTableFromControls(strDataSource As String, clsDatatableControls As DataTable) As Integer
+        Return UpdateTranslationsTable(strDataSource, GetTranslationTextsFromControls(clsDatatableControls))
+    End Function
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>
+    ''' Returns a datatable with all form controls from the list of forms passed
+    ''' </summary>
+    ''' <param name="lstForms">The forms to get the controls for translations</param>
+    ''' <returns>A datatable with 3 columns; form_name, control_name, id_text. </returns>
+    '''--------------------------------------------------------------------------------------------
+    Public Shared Function GetControlsDatatable(lstForms As List(Of Form)) As DataTable
+        Dim clsDatatableControls As New DataTable
+        clsDatatableControls.Columns.Add("form_name", GetType(String))
+        clsDatatableControls.Columns.Add("control_name", GetType(String))
+        clsDatatableControls.Columns.Add("id_text", GetType(String))
+
+        For Each frm As Form In lstForms
+            Dim dctComponents As Dictionary(Of String, Component) = New Dictionary(Of String, Component)
+            clsWinFormsComponents.FillDctComponentsFromControl(frm, dctComponents)
+
+            For Each clsComponent In dctComponents
+                Dim idText As String
+
+                If TypeOf clsComponent.Value Is Control Then
+                    idText = GetActualTranslationText(DirectCast(clsComponent.Value, Control).Text)
+                ElseIf TypeOf clsComponent.Value Is ToolStripItem Then
+                    idText = GetActualTranslationText(DirectCast(clsComponent.Value, ToolStripItem).Text)
+                Else
+                    Throw New Exception("Developer Error: Translation dictionary entry (" & frm.Name & "," & clsComponent.Key & ") contained unexpected value type.")
+                    Exit For
```

Comment:
this line cannot be reached, delete this line?

Reply:
@lloyddewit I wasn't sure about this when I was generalising the code. But I think yes, the component can only be either of the 2 in VB.Net. So I'm deleting it.